Sqlite cleanup + convert to async

brunnre8 opened the pull request:
Tried to clean up the sqlite storage a bit... while working on the msgid thing.

There are no functional changes, other than the fact that the whole message storage interface is now async (has to be, else we can't wait on migrations and so forth)

brunnre8 committed:
sqlite: error if sqlite isn't enabled but search() is called

When we assert that something can't possibly happen, we better
error out rather than jugging on with no error ;)

brunnre8 committed:
sqlite: move export to bottom of the file

This makes it easier to see what's getting exported, rather than
if it's interspersed randomly in the middle of the file

brunnre8 committed:
sqlite: fix docstring

brunnre8 committed:
sqlite: add run helper function

Extract the serialization logic into a single place and
consistently log errors to the console rather than a fire
and forget approach.

brunnre8 committed:
sqlite: create serialize_fetchall helper function

That puts all the serialization logic into one place and
allows us to use async / promises

brunnre8 committed:
sqlite: use serialize_fetchall in getMessages

brunnre8 committed:
sqlite: use serialize_fetchall in search

brunnre8 committed:
extract migrations

brunnre8 committed:
don't call search on a disabled msg provider

A provider might be available, but not functional (broken migration
invalid configuration or what have you).
Don't try to call search in this case.

brunnre8 committed:
sqlite: add serialize_get

brunnre8 committed:
sqlite: convert migrations to async

This removes quite a bunch of indention and callbacks

itsjohncs commented on server/plugins/messageStorage/sqlite.ts:
```diff
@@ -243,31 +227,59 @@ class SqliteMessageStorage implements ISqliteMessageStorage {
+		return this.isEnabled;
+	}
+
+	private run(stmt: string, ...params: any[]) {
+		this.serialize_run(stmt, params).catch((err) =>
+			log.error(`failed to run ${stmt}`, String(err))
+		);
+	}
+
```
Might as well embrace the `async` way fully:

```typescript
private async run(stmt: string, ...params: any[]) {
    try {
        await this.serialize_run(stmt, params)
    } catch (err) {
        log.error(`failed to run ${stmt}`, String(err));
    );
}
```

brunnre8 replied:
yes. That went the way of the Dodo since... just didn't push the new changes yet.

itsjohncs commented on server/plugins/messageStorage/sqlite.ts:
```diff
@@ -243,31 +227,59 @@ class SqliteMessageStorage implements ISqliteMessageStorage {
-				if (err) {
+			this.database.serialize(() => {
+				this.database.run(stmt, params, (err) => {
+					if (!err) {
```
I think this could be a little simpler.

```typescript
if (err) {
    reject(err);
} else {
    resolve();
}
```

brunnre8 replied:
ack, patched in the patchset that will succeed this one

itsjohncs commented on server/plugins/messageStorage/sqlite.ts:
```diff
@@ -162,61 +155,52 @@ class SqliteMessageStorage implements ISqliteMessageStorage {
 		if (!this.isEnabled) {
 			// this should never be hit as messageProvider is checked in client.search()
-			return Promise.resolve([]);
+			throw new Error(
```
I'm nearly always a fan of making failures louder, this seems much better.

itsjohncs reviewed (approved):
LGTM! I didn't pull it down and test it out, but I did a pretty thorough comb through of the code. Including trying to track down what exactly `serialize` is doing ([docs](https://github.com/TryGhost/node-sqlite3/wiki/Control-Flow)).

brunnre8 committed:
messageStorage: convert to async

Message stores are more complicated that a sync "fire and forget"
API allows for.
For starters, non trivial stores (say sqlite) can fail during init
and we want to be able to catch that.
Second, we really need to be able to run migrations and such, which
may block (and fail) the activation of the store.

On the plus side, this pushes error handling to the caller rather
than the stores, which is a good thing as that allows us to eventually
push this to the client in the UI, rather than just logging it in the
server on stdout

brunnre8 commented:
mkey, rewrote the thing, PTAL.

This is not getting any better, so I'd like to avoid adding more things on top until that's figured out

brunnre8 commented:
@itsjohncs please remove the approval / re-ack the PR, I'll ignore it until you actually looked over the changes.
GH doesn't seem to auto revoke the approvals apparently, even though new stuff is pushed

brunnre8 commented on server/helper.ts:
```diff
@@ -183,3 +184,17 @@ function compareWithWildcard(a: string, b: string) {
 	return re.test(b);
 }
+
+function catch_to_error(prefix: string, err: any): Error {
```
That thing is bonkers, but TS yells at me if I shove the error in a string literal from a promise catch block... as that doesn't guarantee that I'm actually looking at an Error (the interface)...

So you need to type assert this and I couldn't figure out how to do it without the code in this function... and I'm not going to inline this in every catch block

itsjohncs reviewed (changes requested):
> please remove the approval / re-ack the PR

dgw commented:
> GH doesn't seem to auto revoke the approvals apparently, even though new stuff is pushed

This is available for branches with Branch Protection turned on. The relevant checkbox is: "Optionally, to dismiss a pull request approval review when a code-modifying commit is pushed to the branch, select **Dismiss stale pull request approvals when new commits are pushed.**"

[More info](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule)

brunnre8 commented:
@itsjohncs / @MaxLeiter mind reviewing this?
I'd like to merge it down with the search fixes.
